### ISSUE TITLE
Flush on every update not working for recorders or collectors 

### ISSUE THREAD

**Issue by bpalmintier** [gridlab-d/gridlab-d#1150]
Initial Checklist prior to submitting the issue. Check the bellow boxes to the best of your knowledge:

- [x] I have not found an open issue describing the same problem.
- [ ] I went through "CONTRIBUTING.MD" before deciding to submit the issue.
- [N/A] While building GridLAB-D, I followed all steps as detailed in GridLAB-D Wiki.

------------------------------------------------------------------------------------------------------------

## Please describe what was the cause of the issue?
recorders and collectors refuse to flush regularly often only after thousands of timesteps and hours of simulation time) despite multiple settings documented as forcing flush on every value update.

This has resulted in significant data loss on long runs, particularly with co-simulation, since a cluster job may timeout before the buffers are ever flushed or otherwise loose lots of simulation time. For example, I have seen no updates for over 7 hours of simulation clock time, or 2-3 hours of wall time.

## Please provide the steps to reproduce.
Current things I've tried:

Setting for flush on every update at the module level:
```
module tape {
    flush_interval -1;
}
```

Setting for flush on every update in individual recorders and collectors. e.g.:
```
object recorder {
     file substation_voltage.csv;
     parent network_node;
     interval 300;
     property voltage_A.real,voltage_A.imag,voltage_B.real,voltage_B.imag,voltage_C.real,voltage_C.imag;
     flush 0;
}

object tape.collector {
    interval 300;
    group "class=triplex_node";
    file triplex_node_stats.csv;
    property min(voltage_12.mag), max(voltage_12.mag), mean(voltage_12.mag), std(voltage_12.mag);
    flush 0;
};
```

Setting for flush when time equals recording interval
```
object recorder {
     file substation_voltage.csv;
     parent network_node;
     interval 300;
     property voltage_A.real,voltage_A.imag,voltage_B.real,voltage_B.imag,voltage_C.real,voltage_C.imag;
     flush 300;
}

object tape.collector {
    interval 300;
    group "class=triplex_node";
    file triplex_node_stats.csv;
    property min(voltage_12.mag), max(voltage_12.mag), mean(voltage_12.mag), std(voltage_12.mag);
    flush 300;
};
```

## Please provide the exact error/warnining/failure message printed on the console.

No Data captured in file. When data is written, it may break in the middle of a line until the next...
...
... (hours pass)
...
update which successfully finishes the line.

## If possible, please provide any screenshots and other supporting information.
I am using the latest release on OSX (pre-compiled from *.dmg) similar behavior exists on the compiled from source version with HELICS support tested under Linux.
(OSX) Version information: GridLAB-D 4.0.0-17448 (Keeler) 64-bit MACOSX RELEASE

## Additional notes
Currently (per documentation) there are two styles of flush control with confusingly opposite parameters:
* `flush` (for recorders and collectors, not working): 0 = every update, -1 = when buffer full
* `flush_interval` (for group_recorders: Working, for tape module: Not working): 0 = when buffer full, -1 = every update
It would be helpful to consolidate to a single semantic approach, although that is lower priority than just getting this feature to work.

-------------------------------------------------------------------------------------------------------------

**Comment by dchassin** [comment]
For one thing I suspect this code isn’t going to work as expected:

my->flush%gl_globalclock==0

**Comment by dchassin** [comment]
It's going to be difficult to reconcile to differing semantics in a way that's backwards compatible.  One thought would be to implement both in both but deprecate one of them.

**Comment by afisher1** [comment]
I'm confused. To me it looks like the ops->flush is just plain wrong for recorder and histogram.

**Comment by afisher1** [comment]
Also, shouldn't the modulus be performed on the global clock relative to the beginning of the simulation? With the suggested change which is correct there would be a bit of a fudge factor on the first flush which is not a big deal I guess.

**Comment by afisher1** [comment]
Agreed

**Comment by bpalmintier** [comment]
Thanks! I'll have to get things set up to compile first, but will try soon-ish. In the mean time I am trying to work around a possible issue around the header not being printed during the last round of recorder runs... it might be related to the flush or not, but will have to play around a bit more before putting in a formal bug report issue.

**Comment by dchassin** [comment]
The fix to recorder.c and collector.c didn't make it through the PR.  See https://github.com/dchassin/gridlabd/pull/5 for details.

**Comment by afisher1** [comment]
@dchassin. My laptop died and I forgot my charger so I edited tape.c, collector.c, and recorder.c on my phone through github and i could only commit a file one at a time that way.

**Comment by dchassin** [comment]
The problem has several causes. See https://github.com/dchassin/gridlabd/commits/aivanova/recorder_flush for more details and some test scripts to verify functionality.  Three problems have been found.

1. The test for triggering a flush in recorder.c and collector.c are not correct.  The terms of the modulo are swapped.

1. The function maps in init.cpp are incorrect for recorder and histogram are linked to collector.

1. tape_file/tape_file.cpp does not export flush_recorder (or flush_collector or flush_histogram).
